Jira Card Plugin

lewtakm opened the pull request:
Jira plugin v0.1.0

Plugin look like this right now:

![chrome_6aqAyQKqwx](https://user-images.githubusercontent.com/36006588/99234075-5e263d80-27f4-11eb-9aba-757cc6239226.png)

After merge main branch needs to be set as default.

dtuite commented:
Thanks Mat. I have a few questions.

 1. What does the Statuses selector do?
 2. Could the component live in an annotation instead of a UI selector?
 3. What would it take to make the activity stream use dark mode? It doesn't fit with the rest of the theme.

lewtakm commented:
> Thanks Mat. I have a few questions.
> 
> 1. What does the Statuses selector do?
> 2. Could the component live in an annotation instead of a UI selector?
> 3. What would it take to make the activity stream use dark mode? It doesn't fit with the rest of the theme.

@dtuite I have updated the PR.
1. Select status of the task for example count only task with status Done, or Done, Backlog etc.
2. I removed this selector.
3. Design updated to adapt to user selected theme. Dark mode:
![8zmeNMhlEu](https://user-images.githubusercontent.com/36006588/99252042-a8b4b380-280e-11eb-901d-a2819f188ece.gif)
Light mode is the same as earlier.

dtuite commented:
Thanks @lewtakm That looks better.

@iain-b could you do a code review please?

iain-b commented on src/api/index.ts:
```diff
@@ -40,8 +43,91 @@ export class JiraAPI {
+    const statusesString = this.convertToString(statusesNames);
+    const jql = `
+      project = "${projectKey}"
+      AND issuetype = "${issueType}"
```
I suppose it's not possible to query like so `issueType in (...)` ?

lewtakm replied:
Actually it is possible but I think operator ```in``` is for multiple options field for example status. I take it directly from JQL generated by Jira itself.

adamdmharvey replied:
FWIW, In Jira Query Language (JQL), `IN` can apply to most types.  In this example, `issueType = "Epic"` is equivalent to `issueType in ("Epic")`.

iain-b replied:
The reason I was asking is that I was hoping we could avoid running multiple JQL queries by returning all the relevant issues. We'd have to calculate counts by issueType client side (unless I'm missing something in jql) but I think that's preferable to running multiple jql queries (in my experience jira can be veeeerrrry slow)

lewtakm replied:
Amazing idea, I will definitely do that

lewtakm replied:
Sadly that was premature, when you create JQL for multiple types it won't return you total number of issues by type. So it won't be possible to merge all of this requests into one.

adamdmharvey replied:
Right! As @iain-b points out, that would need to be calculated client side. So if there were 5 epics, 10 stories, and 3 bugs, your returned issue could would be 18 and you'd need to client-side loop through your result set to do counts.

lewtakm replied:
The problem with that approach is that Jira limits number of elements in response to 100, so for example for a project with 3000 issues I would need to made 30 requests to fetch all of them, add all of them to a single array and then iterate over an array and count issues per type. Using single issueType per query return total number of issues without any limitations. I tested it on my example Jira project. 

To make it more visible lets compare responses size for two types of approaches.
1. Fetch each issueType separately (258 issues in total):
![uAWEpU8](https://user-images.githubusercontent.com/36006588/99702687-6dc4b100-2a96-11eb-87df-2a71724c6d3c.png)
2. Fetch all issuesType together (only first 100 of them, we would need to make 2 more similar requests in order to get all issues):
![zn9GyGk](https://user-images.githubusercontent.com/36006588/99702943-c7c57680-2a96-11eb-8575-4148e95d2ec2.png)

Thanks for pointing this out tho, that was really good idea but sadly Jira REST API is not that friendly as it could be.

iain-b commented on src/components/useRequests.ts:
```diff
@@ -0,0 +1,129 @@
+  message: error?.response?.data?.errorMessages.length && error.response.data.errorMessages[0] || error.request
+});
+
+export const useActivityStream = (size: number) => {
```
In terms of structure could we have a hooks folder alongside components and have each hook in its own file ?

iain-b reviewed (commented):
Very minor, just file structure really and a question about the API
